content/library/advanced-features/forms.md

```diff
@@ -0,0 +1,227 @@
+
+```python
+import streamlit as st
+
+col1,col2 = st.columns([1,2])
+col1.title('Sum:')
+
+with st.form('addition'):
+    a = st.number_input('a')
+    b = st.number_input('b')
+    submit = st.form_submit_button('add')
+
+if submit:
+    col2.title(f'{a+b:.2f}')
+```
+
+<Cloud src="https://doc-forms-process1.streamlit.app/?embed=true" height="360"/>
+
+### Use a callback with session state
+
+You can use a callback to execute a process as a prefix to the script rerunning.
+
+<Important>
+
+When processing newly updated values within a callback, do not pass those values to the callback directly through the `args` or `kwargs` parameters. You need to assign keys to any widget whose value you use within the callback. If you look up the value of that widget from `st.session_state` within the body of the callback, you will be able to access the newly submitted value. See the example below.
+
+</Important>
+
+```python
+import streamlit as st
+
+if 'sum' not in st.session_state:
+    st.session_state.sum = ''
+
+def sum():
+    result = st.session_state.a + st.session_state.b
+    st.session_state.sum = result
+
+col1,col2 = st.columns(2)
+col1.title('Sum:')
+if isinstance(st.session_state.sum, float):
+    col2.title(f'{st.session_state.sum:.2f}')
+
+with st.form('addition'):
+    st.number_input('a', key = 'a')
+    st.number_input('b', key = 'b')
+    st.form_submit_button('add', on_click=sum)
+```
+
+<Cloud src="https://doc-forms-process2.streamlit.app/?embed=true" height="360"/>
+
+### Use `st.experimental_rerun`
+
+If you need to utilize the values from a form before your form in the script, another alternative is using an extra rerun. This can be less resource-efficient though, and may be less desirable that the above options.
+
+```python
+import streamlit as st
+
+if 'sum' not in st.session_state:
+    st.session_state.sum = ''
+
+col1,col2 = st.columns(2)
+col1.title('Sum:')
+if isinstance(st.session_state.sum, float):
+    col2.title(f'{st.session_state.sum:.2f}')
+
+with st.form('addition'):
+    a = st.number_input('a')
+    b = st.number_input('b')
+    submit = st.form_submit_button('add')
+
+# The value of st.session_state.sum is updated at the end of the script rerun,
+# so the displayed value at the top in col2 does not show the new sum. Trigger
+# a second rerun when the form is submitted to update the value above.
+st.session_state.sum = a + b
+if submit:
+    st.experimental_rerun()
+```
+
+<Cloud src="https://doc-forms-process3.streamlit.app/?embed=true" height="360"/>
+
+## Limitations
+
+* An `st.form` cannot be embedded inside another `st.form`.
+* Every form must have an associated `st.form_submit_button`.
+* By definition, `st.button` does not make much sense within a form. Forms are all about batching widget state together, but buttons are inherently stateless. So declaring an st.button inside a form will lead to an error.
```

Comment:
Let's slightly change the "does not make much sense within a form" phrasing here to be a little softer. Might even make sense to link to your buttons guide here.

> So declaring an st.button inside a form will lead to an error.

Unclear what error this might lead to. If the answer is very long, again, linking to the button article where the error is mentioned might help

Reply:
I copied this phrasing from the blog post on forms. I'll test and see how the error is worded.

Reply:
The error generated is absolutely clear and to the point:
> `StreamlitAPIException: st.button() can't be used in an st.form().`

I have reworded the line to match how it is stated in the docstring for clarity:
> `st.button` and `st.download_button` cannot be added to a form.